### ISSUE TITLE
diffusion_distillation colab ( and code ) requires updates to hande deprecations in flax and jax

### ISSUE THREAD

**Issue by xvdp** [google-research/google-research#2976]
It is not possible to run this seminal diffusion distillation project for a series of reasons. 

* flax.optim was deprecated  somewhere between version 4.1 ( which was what i presume the project was built on given the similarity of dates ~ early 2022)
I tried several routes. 

1. build a dockerfile with early 2022 versions python==3.10 jax==0.4.1 flax==0.4.1 & jaxlib==0.4.1 -f https://storage.googleapis.com/jax-releases/jax_releases.html over cuda 11.2 cudnn 8
-- locally i have gpus not tpus, and jax.devices() show no gpus, jaxlib was yanked and the versions there appear under nocuda. So this is a dead end.

2. install those versions in a google colab, but colabs are entrenched on numpy 2.0.2 and new versions of jax flax jaxlib and optax...  even after setting python to 3.10, doing !pip install --force-reinstall --no-cache-dir -r constraints.txt  etc... end up killing the session... 

3. updating model.py to use `optax.adam` or `adamw`  in stead of flax.optim.Adam,  using class `TrainState(flax.training.train_state.TrainState):` in lieu of `@flax.struct.dataclass class TrainState:`
This is the only route that worked, running in TPU on collab it trains the first step, but the reported losses are vastly different (eg. ending in `{'train/gnorm': 801.891845703125, 'train/loss': 3.311255693435669}` instead of `'train/gnorm': 0.6390635371208191, 'train/loss': 0.12706886231899261}` and although converging in the reporting, they dont do so monotonically. I tried removimg weight_decay but same deal.
But even if that works, the training of the student - which is the interesting part, makes use of TrainState, and Im not sure that Im making the new native TrainState structure handle that correctly. So..

A. theres no simple way of resurrecting the old state. 
B theres no simple way to get the same values with the new code, perhaps unless one is familiar with the new TrainState and the paper itself, but I was simply trying to elucidate the distillation steps, that seem to use some shenanigans specifically around that TrainState. 

ITs a pity that projects like this go to rot when they are so cleanly laid out. 

A nice test for updates of colab, jax, jaxlib, optax, flax, and your wonderful Gemini would be to set up a pipeline where results on seminal published projects are maintained.  What better proof that your collective work does indeed work, than to have an AI curator that keeps all code ticking... Or is google structurally too big? It should not be so having one of the prime AI research teams and projects around.
Perhaps you could have internal versions of the original projects running in the original code without having to rely on the oversights of dependency managers, and as new code streams in  your tool would be adapting and translating code, with full system tests. Given that these p4ojects are all well encapsulated (but for the madness of dependency rot) this should be a simple thing to do, no?  On top of which you could leverage Gemini to understand the meaning of passing of time, code rot, versions, and goal oriented programming. Sure it is getting better but it still barfs up code without reasoning on purpose.

But that's not my job i simply was trying to update my knowledge of distillation.

**Comment by xvdp** [comment]
*Edited* I did a partial fix of the deprecation of flax.optim, which appears to fix the distillation, but does not fix training model from scratch. 

I included visualization of the training and distill. Anyone with any working knowledge of jax flax and optax should be able to fix the problem in minutes. It would be valuable to have a working example.

Nice paper. simple and to the point.
